pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/service/SomaticResultUploadService.java

```diff
@@ -0,0 +1,186 @@
+        }
+
+        return new SomaticResultUploadService(
+                signerCredentials,
+                new GoogleBucketClient(projectId, bucketCredentials),
+                realmToUploadBucketMap,
+                cfg.getInt(ApplicationConfigConstants.MAX_SIGN_URL_MIN),
+                new SomaticResultUploadSettings(cfg));
+    }
+
+    public SomaticResultUploadService(ServiceAccountSigner signer, GoogleBucketClient storageClient,
+                                      Map<String, String> realmToUploadBucketMap, int maxSignedUrlMins,
+                                      SomaticResultUploadSettings somaticResultUploadSettings) {
+        this.signer = signer;
+        this.storageClient = storageClient;
+        this.realmToUploadBucketMap = realmToUploadBucketMap;
+        this.maxSignedUrlMins = maxSignedUrlMins;
+        this.somaticUploadSettings = somaticResultUploadSettings;
+    }
+
+    public AuthorizeResult authorizeUpload(String realm, String userId, String ddpParticipantId,
+                                           SomaticResultMetaData somaticResultMetaData) {
+        FileValidator fileValidator = new FileValidator();
+        if (somaticResultMetaData.getFileSize() > somaticUploadSettings.getMaxFileSize()) {
+            return new AuthorizeResult(AuthorizeResultType.FILE_SIZE_EXCEEDS_MAXIMUM, null, null, somaticUploadSettings);
+        }
+        if (!somaticUploadSettings.getMimeTypes().contains(somaticResultMetaData.getMimeType())) {
+            return new AuthorizeResult(AuthorizeResultType.MIME_TYPE_NOT_ALLOWED, null, null, somaticUploadSettings);
+        }
+        if (!fileValidator.isValidFileName("Validating file name",
+                somaticResultMetaData.getFileName(), somaticUploadSettings.getAllowedFileExtensions(), false)) {
+            return new AuthorizeResult(AuthorizeResultType.INVALID_FILE_NAME, null, null, somaticUploadSettings);
+        }
+
+        String fileUUID = UUID.randomUUID().toString();
+        String blobPath = makeBlobPath(somaticResultMetaData, ddpParticipantId, realm, fileUUID);
+        long userIdLong = Long.parseLong(userId);
+        String uploadBucket = getUploadsBucket(realm);
+
+
+        SomaticResultUpload createdUpload = SomaticResultUpload.createFileUpload(realm, ddpParticipantId,
+                somaticResultMetaData.getFileName(), somaticResultMetaData.getMimeType(), uploadBucket, blobPath, userIdLong);
+
+        Map<String, String> headers = Map.of("Content-Type", somaticResultMetaData.getMimeType());
+
+        URL signedURL = storageClient.generateSignedUrl(
+                signer, uploadBucket, blobPath,
+                maxSignedUrlMins, TimeUnit.MINUTES,
+                HttpMethod.PUT, headers);
+
+        return new AuthorizeResult(AuthorizeResultType.OK, signedURL, createdUpload, somaticUploadSettings);
+    }
+
+    public SomaticResultUpload deleteUpload(long userId, int documentId) {
```

Comment:
Are all exceptions that can occur in this method internal errors? Likewise for `authorizeUpload`.

Reply:
No.  Change pushed to clarify which the user can respond to and which are internal.